XBrowser/WebViewController.swift

```diff
@@ -9,57 +9,77 @@ import Combine
 import UIKit
 import WebKit
 
-class WebViewController: UIViewController, WKUIDelegate {
-
-  var viewModel: WebViewModel!
-
-  var webView: WKWebView!
-
-  var cancellables = Set<AnyCancellable>()
-
-  override func loadView() {
-    let webConfiguration = WKWebViewConfiguration()
-    webView = WKWebView(frame: .zero, configuration: webConfiguration)
-    webView.uiDelegate = self
-    webView.allowsBackForwardNavigationGestures = true
-    view = webView
-
-    webView.publisher(for: \WKWebView.canGoBack)
-      .assign(to: \.canGoBack, on: viewModel)
-      .store(in: &cancellables)
+@objc protocol WebViewControllerDelegate {
+    func didGoback()
+    func didGoForward()
+    func goTextUrl(_ url: String? )
+}
 
-    webView.publisher(for: \WKWebView.canGoForward)
-      .assign(to: \.canGoForward, on: viewModel)
-      .store(in: &cancellables)
+class WebViewController: UIViewController, WKUIDelegate, WKNavigationDelegate {
+    
+    var viewModel: WebViewModel!
+    
+    var webView: WKWebView!
+    
+    var cancellables = Set<AnyCancellable>()
+    private var isTapToolbar: Bool = false
+    
+    override func loadView() {
+        let webConfiguration = WKWebViewConfiguration()
+        webView = WKWebView(frame: .zero, configuration: webConfiguration)
+        webView.uiDelegate = self
+        webView.navigationDelegate = self
+        webView.allowsBackForwardNavigationGestures = true
+        view = webView
+        
+        webView.publisher(for: \WKWebView.canGoBack)
+            .assign(to: \.canGoBack, on: viewModel)
+            .store(in: &cancellables)
 
-    webView.publisher(for: \WKWebView.url)
-      .map { $0?.absoluteString }
-      .assign(to: \.url, on: viewModel)
-      .store(in: &cancellables)
+        webView.publisher(for: \WKWebView.canGoForward)
+            .assign(to: \.canGoForward, on: viewModel)
+            .store(in: &cancellables)
+        
+        webView.publisher(for: \WKWebView.url)
+            .map { $0?.absoluteString }
+            .assign(to: \.url, on: viewModel)
+            .store(in: &cancellables)
+        
+        viewModel.$url
+            .sink(receiveValue: { [self] url in
+                guard let url = url,
+                      let _URL = URL(string: url)
+                else { return }
+                let request = URLRequest(url: _URL)
+                if !isTapToolbar {
+                    webView.load(request)
+                }
+            })
+            .store(in: &cancellables)
+    }
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+    }
+}
 
-    viewModel.$url
-      .sink(receiveValue: { [self] url in
+extension WebViewController: WebViewControllerDelegate {
+    func goTextUrl(_ url: String?) {
         guard let url = url,
-          let _URL = URL(string: url)
+              let _URL = URL(string: url)
         else { return }
+        isTapToolbar = true
         let request = URLRequest(url: _URL)
         webView.load(request)
```

Comment:
It's OK to update the web view this way.
But it will be more consistent if the view only reacts/updates on the changes from the view model.

Reply:
Please update your code so that the view only updates when there are changes from the view model!